parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt

```diff
@@ -0,0 +1,121 @@
+package com.parsely.parselyandroid
+
+import android.app.Activity
+import androidx.test.core.app.ActivityScenario
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import java.io.File
+import java.io.FileInputStream
+import java.io.ObjectInputStream
+import java.lang.reflect.Field
+import java.nio.file.Path
+import kotlin.io.path.Path
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.seconds
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withTimeoutOrNull
+import kotlinx.coroutines.yield
+import okhttp3.mockwebserver.MockResponse
+import okhttp3.mockwebserver.MockWebServer
+import okhttp3.mockwebserver.RecordedRequest
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.Test
+import org.junit.runner.RunWith
+
+
+@RunWith(AndroidJUnit4::class)
+class FunctionalTests {
+
+    private lateinit var parselyTracker: ParselyTracker
+    private val server = MockWebServer()
+    private val url = server.url("/").toString()
+    private lateinit var appsFiles: Path
+
+    private fun beforeEach(activity: Activity) {
+        appsFiles = Path(activity.filesDir.path)
+
+        if (File("$appsFiles/$localStorageFileName").exists()) {
+            throw RuntimeException("Local storage file exists. Something went wrong with orchestrating the tests.")
+        }
+    }
+
+    /**
+     * In this scenario, the consumer application tracks more than 50 events-threshold during a flush interval.
+     * The SDK will save the events to disk and send them in the next flush interval.
+     * At the end, when all events are sent, the SDK will delete the content of local storage file.
+     */
+    @Test
+    fun appTracksEventsAboveQueueSizeLimit() {
+        ActivityScenario.launch(SampleActivity::class.java).use { scenario ->
+            scenario.onActivity { activity: Activity ->
+                beforeEach(activity)
+                server.enqueue(MockResponse().setResponseCode(200))
+                parselyTracker = initializeTracker(activity)
+
+                repeat(51) {
+                    parselyTracker.trackPageview("url", null, null, null)
+                }
+            }
+
+            // Waits for the SDK to send events (flush interval passes)
+            val requestPayload = server.takeRequest().toMap()
+            assertThat(requestPayload["events"]).hasSize(51)
+
```

Comment:
Question (❓): Although I understand how this test is working and what it tried to verify, I feel that something is missing here, an assertion of some kind, something that will showcase what was documented on the test case above:

- `The SDK will save the events to disk...`
- `...and send them in the next flush interval.`
- `At the end, when all events are sent...`
- `...the SDK will delete the content of local storage file.`

I mean, can we assert that the SDK is saving the events first? For example, will it save all `51` of them, or only the `1` event that exceed the threshold, something else? Then, what happens during the first flush interval, what happens during the next flush interval, these kind of question. Apologies for trying to push things, I am just trying to review this as a very naive test reader that would like to understand this SDK through testing, assuming I don't have any prior knowledge or the SDK. Wdyt, too much? 🤔

PS: I also naively tried to add something basic, the `assertThat(locallyStoredEvents).isNotEmpty` before the `waitFor { locallyStoredEvents.isEmpty() }`, to see if that will pass, but it failed. 🤷

Reply:
> I mean, can we assert that the SDK is saving the events first?

This is something I pivoted from in 8d4a0d39e04b57a6a05bb260e6978e23905c7c12. IMO, the functional tests shouldn't check for such implementation details. The things we're interested in/testing here are:

- that SDK has sent the same number of events consumer app tracked (asserted in line 65)
- that SDK doesn't store additional data (line 69). Actually, we shouldn't even check this probably in a functional test, but this is a safe-check as I know we have some issues with events duplications.

> Then, what happens during the first flush interval, what happens during the next flush interval, these kind of question.

There is only one flush, and it's tested 👍 I'll be extending this tests suite to cover more cases, like multiple flushes, but here we test "what would happen if user recorded more than 50 events quickly".

> PS: I also naively tried to add something basic, the assertThat(locallyStoredEvents).isNotEmpty before the waitFor { locallyStoredEvents.isEmpty() }, to see if that will pass, but it failed. 🤷

Good point, but it's only timing thing. It seems that assertions above take long enough to give SDK time to clean the file.

Try this:

```diff
Subject: [PATCH] refactor: move the class to Kotlin

This class is relatively simple and difficult to test because of `java.utils.Timer`. That's why I decided to make migration to Kotlin right away, without unit tests coverage first.
---
Index: parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt b/parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt
--- a/parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt	(revision 5900d93eff388bc5b2d98c44072cee07d91b1a96)
+++ b/parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt	(date 1698666962291)
@@ -62,11 +62,11 @@
 
             // Waits for the SDK to send events (flush interval passes)
             val requestPayload = server.takeRequest().toMap()
+            assertThat(locallyStoredEvents).isNotEmpty
             assertThat(requestPayload["events"]).hasSize(51)
 
             // Wait a moment to give SDK time to delete the content of local storage file
             waitFor { locallyStoredEvents.isEmpty() }
-            assertThat(locallyStoredEvents).isEmpty()
         }
     }
 ```

and the test should fail!

WDYT?

Reply:
Thanks for the reply @wzieba ! 👍 

> This is something I pivoted from in https://github.com/Parsely/parsely-android/commit/8d4a0d39e04b57a6a05bb260e6978e23905c7c12. IMO, the functional tests shouldn't check for such implementation details. The things we're interested in/testing here are:
> 
> - that SDK has sent the same number of events consumer app tracked (asserted in line 65)
> - that SDK doesn't store additional data (line 69). Actually, we shouldn't even check this probably in a functional test, but this is a safe-check as I know we have some issues with events duplications.

👍 if you are also 👍 with that, I just wanted to question it a bit... 😊

> There is only one flush, and it's tested 👍 I'll be extending this tests suite to cover more cases, like multiple flushes, but here we test "what would happen if user recorded more than 50 events quickly".

Thanks for the clarification on that. 👍 

> Good point, but it's only timing thing. It seems that assertions above take long enough to give SDK time to clean the file.
> 
> Try this:
> 
> + assertThat(locallyStoredEvents).isNotEmpty
> 
> and the test should fail!
> 
> WDYT?

Yea, it is hard to assert those things... 😭 

FYI: The test should fail, I don't understand that, the test should success with this change, is it not? It failing was the actual problem I had as well, as I would have expected it to succeed before waiting... 🤔

PS: Apologies, I am a bit confused it seems... 🤷 

> - assertThat(locallyStoredEvents).isEmpty()

Why did you remove this assertion from the patch you shared? 🤔

Reply:
Yes, sorry I made some confusion with the diff! Please ignore the diff in previous comment. But it's still case of timing - check this:
```diff
Subject: [PATCH] fix: add `intern` to `ROOT_URL` of emulator localhost
---
Index: parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt b/parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt
--- a/parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt	(revision d2a25dbc864396d67dc517943a5a079cffb86304)
+++ b/parsely/src/androidTest/java/com/parsely/parselyandroid/FunctionalTests.kt	(date 1698668794333)
@@ -71,6 +71,7 @@
     }
 
     private fun RecordedRequest.toMap(): Map<String, List<Event>> {
+        assertThat(locallyStoredEvents).isEmpty()
         val listType: TypeReference<Map<String, List<Event>>> =
             object : TypeReference<Map<String, List<Event>>>() {}
 

```

In this case the test **will** fail although, in theory, it shouldn't because after making the request, SDK will clean the local file. But, as the things happen on different threads, we don't have certainty what will happen first - assertion, or file deletion.

In case of this test - the assertions **and** JSON mapping take enough time, that, at least in theory, we shouldn't need to wait. Yet, relying on "JSON serialization being slow" as a determiner of test success is super flaky. Hence, the `waitFor` method.

Reply:
> But it's still case of timing...

Thanks for sharing a new diff, and I see, thanks for explaining again! ✅ 

I don't want to block this PR, feel free to merge it when you think it is ready, I just wanted to question it a bit just to get a better idea on the actual value of this functional test and how you would that be of use later on. 👍

Reply:
Thanks for the questions! We can always iterate over how those tests look now. For now, I think this test brings value - we're sure that saving and getting data from the local file is working fine and doesn't produce duplicates.

Reply:
👍 💯 🥇